addons/controllers/tanzuclusterbootstrap_controller.go

```diff
@@ -0,0 +1,586 @@
+	tanzuClusterBootstrap, err := r.createOrPatchTanzuClusterBootstrapFromTemplate(cluster, log)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if tanzuClusterBootstrap == nil {
+		return ctrl.Result{}, nil
+	}
+
+	remoteClient, err := util.GetClusterClient(r.context, r.Client, r.Scheme, clusterapiutil.ObjectKey(cluster))
+	if err != nil {
+		log.Error(err, "Error getting remote cluster client")
+		return ctrl.Result{}, err
+	}
+	// TODO handle kapp as a remote packageinstall from mgmt cluster
+	// handle cni, cpi, csi
+	// https://github.com/vmware-tanzu/tanzu-framework/issues/1587
+
+	// handle additionalPackages
+	for _, additionalPkg := range tanzuClusterBootstrap.Spec.AdditionalPackages {
+		// TODO packageinstall (rbac, serviceaccount, package, packageinstall)
+		// https://github.com/vmware-tanzu/tanzu-framework/issues/1589
+		secret, err := r.createOrPatchPackageInstallSecret(cluster, additionalPkg, remoteClient, tanzuClusterBootstrap.Namespace, log)
+		if err != nil {
+			log.Error(err, "failed to createOrPatchPackageInstallSecret")
+			return ctrl.Result{}, err
+		}
+		if secret != nil {
+			log.Info("created secret for package in cluster", "secret", secret)
+		}
+		// set watches on provider objects in additional packages if not already set
+		if additionalPkg.ValuesFrom.ProviderRef != nil {
+			if err := r.watchProvider(additionalPkg.ValuesFrom.ProviderRef, tanzuClusterBootstrap.Namespace, log); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// createOrPatchTanzuClusterBootstrapFromTemplate will get, clone or update a TanzuClusterBootstrap associated with a cluster
+// all linked secret refs and object refs are cloned into the same namespace as TanzuClusterBootstrap
+func (r *TanzuClusterBootstrapReconciler) createOrPatchTanzuClusterBootstrapFromTemplate(cluster *clusterapiv1beta1.Cluster,
+	log logr.Logger) (*runtanzuv1alpha3.TanzuClusterBootstrap, error) {
+
+	tkrName := util.GetTKRNameForCluster(r.context, r.Client, cluster)
+	if tkrName == "" {
+		log.Info("cluster does not have an associated TKR")
+		return nil, nil
+	}
+
+	tanzuClusterBootstrapTemplate := &runtanzuv1alpha3.TanzuClusterBootstrapTemplate{}
+	key := client.ObjectKey{Namespace: constants.TKGSystemNS, Name: tkrName}
+	if err := r.Client.Get(r.context, key, tanzuClusterBootstrapTemplate); err != nil {
+		log.Error(err, "unable to fetch tanzuClusterBootstrapTemplate", "objectkey", key)
+		return nil, err
+	}
+
+	tanzuClusterBootstrap := &runtanzuv1alpha3.TanzuClusterBootstrap{}
+	err := r.Client.Get(r.context, client.ObjectKeyFromObject(cluster), tanzuClusterBootstrap)
+	// if found and resolved tkr is the same, return found object as the TKR is supposed to be immutable
+	// also preserves any user changes
+	if err == nil && tkrName == tanzuClusterBootstrap.Status.ResolvedTKR {
+		return tanzuClusterBootstrap, nil
+	}
+
+	if !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+	if tanzuClusterBootstrap.UID == "" {
```

Comment:
question: will this condition always be True ?

Reply:
True when object is not found on line 212

Reply:
is it equivalent to ?
```
if !apierrors.IsNotFound(err) {
    return nil, err
} else {
    log.Info("tanzuClusterBootstrap for cluster does not exist, cloning from template")
    ......

}
```

Reply:
probably the same but our linters complain